### PR TITLE
Match exact host in autoplay whitelist before matching eTLD+1

### PR DESCRIPTION
**Pull request by pilgrim-brave**
Resolves https://github.com/brave/brave-browser/issues/4621

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_shields/browser/autoplay_whitelist_service.cc [hunk 1]
```diff
@@ -32,6 +32,12 @@ AutoplayWhitelistService::~AutoplayWhitelistService() {
 
 bool AutoplayWhitelistService::ShouldAllowAutoplay(const GURL& url) {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  // look for exact host match (this is the only way that subdomains
+  // listed in the autoplay whitelist will match)
+  if (url.has_host() &&
```
Comment (bbondy) [inline components/brave_shields/browser/autoplay_whitelist_service.cc]: looks right to me, but will wait for confirmation that it passes.  A couple of unit tests would be super too.

Comment (pilgrim-brave) [reply components/brave_shields/browser/autoplay_whitelist_service.cc]: simplified the test data file and added multiple tests

**Comment by pilgrim-brave** [comment]
AFAICT this only failed because of Github's connectivity issues yesterday, but I'll rebase it and run it through again just to be sure.

**Comment by bsclifton** [comment]
Added missing milestone to this PR and https://github.com/brave/brave-browser/issues/4621

**Comment by pilgrim-brave** [comment]
uplifted to release

**Comment by kjozwiak** [comment]
Reproduced the original issue on `macOS 10.15.3 x64` using the following builds:

```
Brave | 1.7.21 Chromium: 80.0.3987.122 (Official Build) nightly (64-bit)
--- | ---
Revision | cf72c4c4f7db75bc3da689cd76513962d31c7b52-refs/branch-heads/3987@{#943}
OS | macOS Version 10.15.3 (Build 19D76)
```

```
Brave | 1.4.95 Chromium: 80.0.3987.122 (Official Build) (64-bit)
-- | --
Revision | cf72c4c4f7db75bc3da689cd76513962d31c7b52-refs/branch-heads/3987@{#943}
OS | macOS Version 10.15.3 (Build 19D76)
```

* Using the STR mentioned via https://github.com/brave/brave-browser/issues/4621#issue-449129909, reproduced the issue of the audio not working once pressing on `Test` under `Speaker`.

Screen Shot 2020-03-03 at 6 39 13 PM

Verification **PASSED** on `macOS 10.15.3 x64` using the following build:

```
Brave | 1.7.35 Chromium: 80.0.3987.122 (Official Build) nightly (64-bit)
-- | --
Revision | cf72c4c4f7db75bc3da689cd76513962d31c7b52-refs/branch-heads/3987@{#943}
OS | macOS Version 10.15.3 (Build 19D76)
```

* Went through the STR outlined via https://github.com/brave/brave-browser/issues/4621#issue-449129909 and ensured that sound was being played when selecting `Test` under `Speaker`.
* Had a quick call with @rebron and ensured that the sound was being played without the need to allow autoplay permissions under https://meet.google.com.

**Comment by kjozwiak** [comment]
Moving this out of https://github.com/brave/brave-core/milestone/56 as this specific PR is on `master`.  https://github.com/brave/brave-core/pull/4788 is the PR that belongs in https://github.com/brave/brave-core/milestone/56.

**Pull request merged**